Generalize IValue's aliased hash handling for opaque tensors

cbalioglu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70383
* #70380
* __->__ #70371

This PR generalizes the aliased hash handling for opaque tensors beyond MKL-DNN.

Differential Revision: [D33301787](https://our.internmc.facebook.com/intern/diff/D33301787/)

cbalioglu committed:
Generalize IValue's aliased hash handling for opaque tensors

This PR generalizes the aliased hash handling for opaque tensors beyond MKL-DNN.

Differential Revision: [D33301787](https://our.internmc.facebook.com/intern/diff/D33301787/)

[ghstack-poisoned]

cbalioglu committed:
Update on "Generalize IValue's aliased hash handling for opaque tensors"


This PR generalizes the aliased hash handling for opaque tensors beyond MKL-DNN.

Differential Revision: [D33301787](https://our.internmc.facebook.com/intern/diff/D33301787/)

[ghstack-poisoned]

cbalioglu committed:
Update on "Generalize IValue's aliased hash handling for opaque tensors"


This PR generalizes the aliased hash handling for opaque tensors beyond MKL-DNN.

Differential Revision: [D33301787](https://our.internmc.facebook.com/intern/diff/D33301787/)

[ghstack-poisoned]

cbalioglu committed:
Update on "Generalize IValue's aliased hash handling for opaque tensors"


This PR generalizes the aliased hash handling for opaque tensors beyond MKL-DNN.

Differential Revision: [D33301787](https://our.internmc.facebook.com/intern/diff/D33301787/)

[ghstack-poisoned]

cbalioglu committed:
Update on "Generalize IValue's aliased hash handling for opaque tensors"


This PR generalizes the aliased hash handling for opaque tensors beyond MKL-DNN.

Differential Revision: [D33301787](https://our.internmc.facebook.com/intern/diff/D33301787/)

[ghstack-poisoned]

zou3519 commented on aten/src/ATen/core/ivalue.h:
```diff
@@ -309,6 +303,13 @@ struct TORCH_API IValue final {
              isAliasOf(a, b.col_indices());
     }
 
+    // Opaque tensors such as the ones constructed by the MKL-DNN backend
+    // don't have storage so we just compare their TensorImpls.
+    // TODO: Find way to expose alias info for opaque tensors.
+    if (!a.has_storage() || !b.has_storage()) {
+      return a.unsafeGetTensorImpl() == b.unsafeGetTensorImpl();
+    }
```
I'm trying to understand the implications of this. @bdhirsh are there other tensors other than MKLDNN and the functorch ones (e.g. BatchedTensor) that don't have storage?

The other question is probably for the JIT folks (@suo, @eellison) -- do you know why IValue has isAliasOf and what it's used for?

bdhirsh replied:
From a quick look around, the only others I know are `SparseTensorImpl` and `UndefinedTensorImpl`. Meta tensors actually do have storages, since they're useful for tracking alias info.

cbalioglu replied:
I would add to that any 3rd-party tensor type that uses `OpaqueTensorImpl`.

eellison replied:
It's used in 
a) alias schema testing
b) in freezing, so that don't make a parameter a constant if it is aliased elsewhere in the module and written to.

zou3519 replied:
Okay if the only place this is used is JIT freezing then this seems fine

cbalioglu committed:
Update on "Generalize IValue's aliased hash handling for opaque tensors"


This PR generalizes the aliased hash handling for opaque tensors beyond MKL-DNN.

Differential Revision: [D33301787](https://our.internmc.facebook.com/intern/diff/D33301787/)

[ghstack-poisoned]
